### PR TITLE
Fix ios build error when use_frameworks is on and fabric is off

### PR DESCRIPTION
**Pull request by Kudo**
## Summary

when `use_frameworks!` is on, there are errors like:
```
'FBReactNativeSpec/FBReactNativeSpec.h' file not found
#import <FBReactNativeSpec/FBReactNativeSpec.h>
```
this error may come from from https://github.com/facebook/react-native/commit/f7e4c07c84b6 regression.

when `use_frameworks!` is on, xcode will search headers from framework directories, the correct imports would be `#import <React_Codegen/FBReactNativeSpec/FBReactNativeSpec.h>` (xcode will transform dash to underscore, so it is `React_Codegen` but not `React-Codegen`). in the other hand, when `use_frameworks!` is off, the correct import is `#import <React-Codegen/FBReactNativeSpec/FBReactNativeSpec.h>`.

~to make it compatible for both modes, i would propose to use a valid module name for the codegen pod. that is the `ReactCodegen`. supposedly can define different module name in podspecs either by [`header-dir`](https://guides.cocoapods.org/syntax/podspec.html#header_dir) or [`module_name`](https://guides.cocoapods.org/syntax/podspec.html#module_name). however, i don't recommend to use these as [it has a known issue when DEFINES_MODULE=YES](https://github.com/CocoaPods/CocoaPods/issues/11136). that's why i use the `ReactCodegen` name.~
Update: to make it compatible for both modes, i merged `FBReactNativeSpec` and `React-Codegen`. the final name is `FBReactNativeSpec.podspec`, so the final valid import will be `#import <FBReactNativeSpec/FBReactNativeSpec/FBReactNativeSpec.h>`. this is specific for `header_mappings_dir = '.'` mode. to keep the import as original `#import <FBReactNativeSpec/FBReactNativeSpec.h>`, i removed `header_mappings_dir` in non-fabric mode. the workaround has a downside where assuming there're no other codegen libraries. otherwise, it may also break imports, e.g. `#import <ScreenshotmanagerSpec/ScreenshotmanagerSpec.h> -> #import <FBReactNativeSpec/ScreenshotmanagerSpec.h>`.

this fix is specific for old architecture (fabric is off).

when fabric is on, there are other errors from duplicated headers when copying to build folder. [the reason is that framework build would try to flatten headers](https://mkonrad.net/2015/03/29/xcode-static-libraries-preserving-header-directory-structure.html). we have `primitives.h` in different folders and they would be flattened into `React_Fabric.framework/Headers`. to be honest, i don't know how to deal with the problem in the meantime,  maybe subspecs are not enough, we should separate them from subspecs to dedicated podspecs so that we can have these targets as different frameworks.

## Changelog

[iOS] [Fixed] - Fix iOS build error when Podfile `use_frameworks!` is on and Fabric is off

## Test Plan

verify with rn-tester
1. change `fabric_enabled` to false in `packages/rn-tester/Podfile`
2. `USE_FRAMEWORKS=1 pod install`
3. build rn-tester in xcode

### CODE REVIEW

**Comment by mikehardy** [comment]
@Kudo this may be an odd request and/or difficult - if so please ignore - but if you still have this file tree open + clean somewhere, I'd love a diff of this so I could apply locally via patch-package and test in my reproduction of the same issue, then I could confirm it works as well

If that's not easy for any reason just let me know, the changes are pretty easy and though it would be slightly more error-prone than a diff / patch-package I could apply them locally as well for testing + confirmation

**Comment by Kudo** [comment]
@mikehardy thanks for your help. [this commit patch](https://github.com/facebook/react-native/commit/ae024aff85688b1c70b2569774be187dd374a6c0.patch) should be good enough for you to verify.

**Comment by mikehardy** [comment]
Oh of course github generates them :facepalm: :laughing:

**Comment by mikehardy** [comment]
Very tiny differences between the github PR patch and what gets installed in an app, but nothing that should alter validity of test results, here's what it looked like after pulling the patch down and applying it manually inside node_modules against react-native 0.68.0-rc.2:

```

mike@osxvm-little:~/work/Invertase/rnfbdemo/rnfbdemo/node_modules/react-native (rn68) % patch -p1 < ../../../react-native-ios-static-compile-fix-PR33379.patch 
patching file Libraries/Blob/RCTBlobManager.mm
patching file Libraries/Blob/RCTFileReaderModule.mm
patching file Libraries/Blob/React-RCTBlob.podspec
patching file Libraries/Image/RCTImageEditingManager.mm
patching file Libraries/Image/RCTImageLoader.mm
patching file Libraries/Image/RCTImageStoreManager.mm
patching file Libraries/Image/React-RCTImage.podspec
patching file Libraries/LinkingIOS/RCTLinkingManager.mm
patching file Libraries/LinkingIOS/React-RCTLinking.podspec
patching file Libraries/NativeAnimation/RCTNativeAnimatedModule.mm
patching file Libraries/NativeAnimation/RCTNativeAnimatedTurboModule.mm
patching file Libraries/NativeAnimation/React-RCTAnimation.podspec
patching file Libraries/Network/RCTNetworking.mm
patching file Libraries/Network/React-RCTNetwork.podspec
patching file Libraries/PushNotificationIOS/RCTPushNotificationManager.mm
patching file Libraries/PushNotificationIOS/React-RCTPushNotification.podspec
patching file Libraries/Settings/RCTSettingsManager.mm
patching file Libraries/Settings/React-RCTSettings.podspec
patching file Libraries/Vibration/RCTVibration.mm
patching file Libraries/Vibration/React-RCTVibration.podspec
patching file React/CoreModules/RCTAccessibilityManager.mm
patching file React/CoreModules/RCTActionSheetManager.mm
patching file React/CoreModules/RCTAlertManager.mm
patching file React/CoreModules/RCTAppState.mm
patching file React/CoreModules/RCTAppearance.mm
patching file React/CoreModules/RCTAsyncLocalStorage.mm
patching file React/CoreModules/RCTClipboard.mm
patching file React/CoreModules/RCTDevLoadingView.mm
patching file React/CoreModules/RCTDevMenu.mm
patching file React/CoreModules/RCTDevSettings.mm
patching file React/CoreModules/RCTDevSplitBundleLoader.mm
patching file React/CoreModules/RCTDeviceInfo.mm
patching file React/CoreModules/RCTExceptionsManager.mm
patching file React/CoreModules/RCTI18nManager.mm
patching file React/CoreModules/RCTKeyboardObserver.mm
patching file React/CoreModules/RCTLogBox.mm
patching file React/CoreModules/RCTPlatform.mm
patching file React/CoreModules/RCTRedBox.mm
patching file React/CoreModules/RCTSourceCode.mm
patching file React/CoreModules/RCTStatusBarManager.mm
patching file React/CoreModules/RCTTiming.mm
patching file React/CoreModules/RCTWebSocketModule.mm
patching file React/CoreModules/React-CoreModules.podspec
patching file React/React-RCTFabric.podspec
can't find file to patch at input line 676
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/packages/rn-tester/NativeComponentExample/MyNativeView.podspec b/packages/rn-tester/NativeComponentExample/MyNativeView.podspec
|index 7e0fe5281bac..c7c5b3c87a71 100644
|--- a/packages/rn-tester/NativeComponentExample/MyNativeView.podspec
|+++ b/packages/rn-tester/NativeComponentExample/MyNativeView.podspec
--------------------------
File to patch: 
Skip this patch? [y] y
Skipping patch.
1 out of 1 hunk ignored
patching file scripts/react_native_pods.rb
Hunk #2 succeeded at 343 (offset -6 lines).
Hunk #3 succeeded at 360 (offset -6 lines).
Hunk #4 succeeded at 385 (offset -6 lines).
Hunk #5 succeeded at 466 (offset -6 lines).
Hunk #6 succeeded at 510 (offset -6 lines).
mike@osxvm-little:~/work/Invertase/rnfbdemo/rnfbdemo/node_modules/react-native (rn68) % 
```
Resulting patch file (for use with patch-package) for anyone else that wants to test is
[react-native+0.68.0-rc.2.patch.gz](https://github.com/facebook/react-native/files/8186798/react-native%2B0.68.0-rc.2.patch.gz)
 
With the patch integrated, my static frameworks build test now passes!
Reproducible demonstration of compile failure and compile working is present in commit history here https://github.com/mikehardy/rnfbdemo/commits/rn68/make-demo.sh

**Review by mikehardy — APPROVED** [review]
I tested this change via the patch I attached in comment on the PR, for these scenarios, they all worked:

- fabric disabled, hermes/flipper enabled, release mode
- fabric disabled, hermes/flipper enabled, debug mode
- fabric disabled, hermes/flipper disabled, static frameworks, debug mode

The matrix of possibilities for compiles is pretty big at this point now that I think about it! But these are my normal cases and represent the normal cases for react-native-firebase users so they are what I test.

All works, and the PR looks reasonable

**Comment by mikehardy** [comment]
Just wanted to add an update that I tested the same set of combinations but on an M1 mac as well (previous comment was an intel mac, sorry I did not specify), and everything worked on it as well, same :heavy_check_mark:  results.
Let me know if there is anything more I can do.

**Comment by Kudo** [comment]
i am going to rebase main again to fix the conflicts.

**Comment by mikehardy** [comment]
I suppose this is going to be tricky to apply to 0.68 then, because your previous diff applied against the 0.68 branch - if there were conflicts with main it may require 2 PRs ? (one against main, one against 68 branch?). It definitely worked against 0.68 though, I should be specific and say all my testing was applying diff via patch tools against 0.68.0-rc.2

**Comment by Kudo** [comment]
the conflict is just one line difference introduced from https://github.com/facebook/react-native/commit/c2e4ae39b8a5c6534a3fa4dae4130166eda15169. i doubt the pr may break react-native internal BUCK builder because the `#import <FBReactNativeSpec/FBReactNativeSpec.h> -> #import <ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h>` change. let's review this on main first. once it landed, we can cherry pick to 0.68-stable.

File: React/CoreModules/RCTSourceCode.mm
Comment (cortinico) [inline React/CoreModules/RCTSourceCode.mm:10]: Commenting here but this applies to the whole PR.

Sadly, we can't merge this as it is, and we have to find a different approach to this problem. Specifically this is breaking our internal setup (i.e. building React Native iOS with Buck, which sadly we don't use is OSS).

Specifically here we fail internally with:

```
xplat/js/react-native-github/React/CoreModules/RCTSourceCode.mm:10:9: fatal error: 'ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h' file not found
#import <ReactCodegen/FBReactNativeSpec/FBReactNativeSpec.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Review by cortinico — CHANGES_REQUESTED** [review]

File: React/CoreModules/RCTSourceCode.mm
Comment (Kudo) [reply React/CoreModules/RCTSourceCode.mm:10]: sure, i believed this change would break Buck builder 🤔 could you share more insight for how the Buck targets organized internally? from podspec of view, that's be little weird for `FBReactNativeSpec.podspec` doesn't have any source_files and `React-Codegen` puts source_files to `FBReactNativeSpec` import namespace. is it possible to merge these two?

Comment (cortinico) [reply React/CoreModules/RCTSourceCode.mm:10]: So ideally we can't change those `#import <FBReactNativeSpec/FBReactNativeSpec.h>` statements as our internal builds rely on them. I'd try to tweak the Podspecs as much as possible to tackle this.

Sadly this is a bigger problem, that is affecting how we do iOS in OSS, and we should look into this problem from a more holistic point of view.

For the sake of 0.68.x, if we could find a patch on the Podspec level, that would be awesome.

cc @cipolleschi @dmitryrykun for visiblity.

Comment (cortinico) [reply React/CoreModules/RCTSourceCode.mm:10]: So ideally we can't change those `#import <FBReactNativeSpec/FBReactNativeSpec.h>` statements as our internal builds rely on them. I'd try to tweak the Podspecs as much as possible to tackle this.

Sadly this is a bigger problem, that is affecting how we do iOS in OSS, and we should look into this problem from a more holistic point of view.

For the sake of 0.68.x, if we could find a patch on the Podspec level, that would be awesome.

cc @cipolleschi @dmitryrykun for visiblity.

Comment (Kudo) [reply React/CoreModules/RCTSourceCode.mm]: updated the pr to keep `#import <FBReactNativeSpec/FBReactNativeSpec.h> ` imports. please take a look when you get a chance.

this diff is for 0.68.0-rc.2 in case someone need to verify: https://gist.github.com/Kudo/17dde98d7721601697d2dc05ec1dcba1

**Comment by cglacet** [comment]
Hello, I'm having the error you described here (react-native 0.70.6), the path problem seems to arise in multiple locations for me:

```
▸ Compiling RCTVibration.mm

❌  /Users/cglacet/Kune/Clients/client/node_modules/react-native/Libraries/Vibration/RCTVibration.mm:11:9: 'FBReactNativeSpec/FBReactNativeSpec.h' file not found

#import <FBReactNativeSpec/FBReactNativeSpec.h>
                                        
▸ Compiling RCTNetworking.mm

❌  /Users/cglacet/Kune/Clients/client/node_modules/react-native/Libraries/Network/RCTNetworking.mm:11:9: 'FBReactNativeSpec/FBReactNativeSpec.h' file not found

#import <FBReactNativeSpec/FBReactNativeSpec.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

▸ Compiling RCTSettingsManager.mm

❌  /Users/cglacet/Kune/Clients/client/node_modules/react-native/Libraries/Settings/RCTSettingsManager.mm:10:9: 'FBReactNativeSpec/FBReactNativeSpec.h' file not found

#import <FBReactNativeSpec/FBReactNativeSpec.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

▸ Compiling RCTNativeAnimatedTurboModule.mm

❌  /Users/cglacet/Kune/Clients/client/node_modules/react-native/Libraries/NativeAnimation/RCTNativeAnimatedTurboModule.mm:8:9: 'FBReactNativeSpec/FBReactNativeSpec.h' file not found

#import <FBReactNativeSpec/FBReactNativeSpec.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

▸ Compiling RCTNativeAnimatedModule.mm

❌  /Users/cglacet/Kune/Clients/client/node_modules/react-native/Libraries/NativeAnimation/RCTNativeAnimatedModule.mm:8:9: 'FBReactNativeSpec/FBReactNativeSpec.h' file not found

#import <FBReactNativeSpec/FBReactNativeSpec.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ```
 
 I searched for this file and it only appears in several `./ios/Build` subdirectories, for example this one:

```bash
./ios/Build/generated/ios/FBReactNativeSpec/FBReactNativeSpec.h
```

I thought maybe I should remove the build directory just in case, but it didn't work either and I still can't compile `RCTNativeAnimatedModule`, `RCTNativeAnimatedTurboModule` and `RCTSettingsManager`. I've also tried removing `node_modules`, `Pods` and both `yarn.lock` and `Podfile.lock` without success:

```bash
❯ rm -rf yarn.lock node_modules ios/Pods ios/Podfile.lock ios/Build &&
yarn install && cd ios && pod install && .. && yarn ios:build
```

But it didn't solve the issue.

Here is my Podfile content: 

```
# Manual merge from : https://raw.githubusercontent.com/react-native-community/rn-diff-purge/release/0.67.2/RnDiffApp/ios/Podfile
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '12.4'
install! 'cocoapods', :deterministic_uuids => false

$RNFirebaseAsStaticFramework = true
$RNMapboxMapsImpl = 'mapbox'

project 'Kune', 
  'Release' => :release, 
  'Staging.Release' => :release, 
  'Development.Release' => :release, 
  'Debug' => :debug,
  'Staging.Debug' => :debug,
  'Development.Debug' => :debug


use_frameworks! :linkage => :static

target 'Kune' do
  config = use_native_modules!(packages_to_skip: ['react-native-flipper'])
  pod 'react-native-config', :path => '../node_modules/react-native-config'
  flags = get_default_flags()

  use_react_native!(
    :path => config[:reactNativePath],
    :hermes_enabled => true,
    :fabric_enabled => flags[:fabric_enabled],
    :flipper_configuration => FlipperConfiguration.disabled,
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'KuneTests' do
    inherit! :complete
  end

  pre_install do |installer|
    $RNMBGL.pre_install(installer)
  end

  post_install do |installer|
    react_native_post_install(
      installer,
      :mac_catalyst_enabled => false
    )
    installer.pods_project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings.delete 'IPHONEOS_DEPLOYMENT_TARGET'
      end
    end
    # Ref: https://stackoverflow.com/a/63955114/1720199
    installer.pods_project.build_configurations.each do |config|
      config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
    end
    $RNMBGL.post_install(installer)
  end

  pod 'React-Core/RCTWebSocket', :path => '../node_modules/react-native/'
  pod 'Firebase', :modular_headers => true
  pod 'FirebaseCore', :modular_headers => true
  pod 'GoogleUtilities', :modular_headers => true
  pod 'Firebase/Analytics'
  pod 'Firebase/Auth'
  pod 'GoogleSignIn'
end
```

I couldn't find any help about this, sorry for asking this here, but you seem to understand the issue so I guessed maybe there is something obvious I missed. 

Thanks.

**Pull request closed without merging**
